numba/cuda/simulator/cudadrv/devicearray.py

```diff
@@ -218,7 +218,30 @@ def device_array(*args, **kwargs):
 
 
 def device_array_like(ary, stream=0):
-    return FakeCUDAArray(np.empty_like(ary))
+    # Avoid attempting to recompute strides if the default strides will be
+    # sufficient to create a contiguous array.
+    if ary.is_c_contiguous() or ary.ndim <= 1:
+        return FakeCUDAArray(np.ndarray(shape=ary.shape, dtype=ary.dtype))
+
+    # Otherwise, we need to compute new strides using an algorithm adapted from
+    # Numpy's PyArray_NewLikeArrayWithShape in core/src/multiarray/ctors.c. We
+    # permute the strides in ascending order then compute the stride for the
+    # dimensions with the same permutation.
+
+    # Stride permuation. E.g. a stride array (4, -2, 12) becomes
+    # [(1, -2), (0, 4), (2, 12)]
+    strideperm = [ x for x in enumerate(ary.strides) ]
+    strideperm.sort(key = lambda x: x[1])
+
+    # Compute new strides using permutation
+    strides = [0] * len(ary.strides)
+    stride = ary.dtype.itemsize
+    for i_perm, _ in strideperm:
+        strides[i_perm] = stride
+        stride *= ary.shape[i_perm]
+    strides = tuple(strides)
```

Comment:
Not for now, but one day it'd be nice to have a "common" or "shared" details module from which the real the the fake device array inherits/is composed.

Reply:
I agree - opened Issue #5052.

Reply:
Great, thanks.